### ISSUE TITLE
The consensus took 65 rounds

### ISSUE THREAD

**Issue by xieyiwen** [tendermint/tendermint#2251]
**Tendermint version 0.23.0**  

**Environment**:
- **ubuntu X64**
    2 vCPU 2 GB 1Mbps
 
- **docker**:

**What happened**:
The consensus too slow, consensus took 65 rounds。
the block is too much, but i found many rounds have enough votes. it should complete soon. 

**What you expected to happen**:
Complete the consensus no more than 10 rounds

**Have you tried the latest version**: yes

**How to reproduce it** (as minimally and precisely as possible):
I send  tx/s by 50 go thread.  one Tx include 17054 chars.
the node's mempool cache 259 txs

**Logs (paste a small part showing an error (< 10 lines) or link a pastebin, gist, etc. containing more of the log file)**:

The big blockInfo
```
 Block{
  Header{
    ChainID:        test-chain-ybl
    Height:         14043
    Time:           2018-08-21 06:56:07.801083933 +0000 UTC
    NumTxs:         259
    TotalTxs:       266
    LastBlockID:    7197306281BBB6466B76926C68250195D9BDB014:1:29B22A6A1491
    LastCommit:     50B2D83F6463AD406C28CAF6D1D0AEF6DCCF47D6
    Data:           3E2838E43D52804F9A6AC587F2502FE6FB4927DA
    Validators:     6F56BED5F521BF1CB38E9F4B5D3BA5534BCBE834
    App:            7197306281BBB6466B76926C68250195D9BDB014
    Consensus:       D6B74BB35BDFFD8392340F2A379173548AE188FE
    Results:        30FA030E85DCB8EDC9B5F519C10DEFD1032246A8
    Evidence:
  }#5DC7A821037E7482161DC0378444F98BA82D315D
  Data{
    EEBA64208E23746C8921B6377ADA3B59ED3BA015 (17048 bytes)
    55534358920B20AE4AC98781C2996FA162E36C59 (17034 bytes)
    C59A676AFF575461F815D0EED2562AACA30DC7FF (17028 bytes)
    F53E64BFBE69D3C0EFECD557BBF974997D88E22E (17040 bytes)
    9283A8E0FC348BD16CB8F142AE55000B4AA42EBA (17036 bytes)
    6E54B0658006C5C0871E75781ABBECEC65B4FDFF (17020 bytes)
    019A19A28FB1DA3ED4F00D97AFB4D958CCE74114 (17042 bytes)
    8EB8B2D654CAFC09FE3D65EF1EF69B2BA91FED63 (17050 bytes)
    086CD3ACDD8CC6628BC08DC4B3EA96DF2ADDB5B5 (17026 bytes)
    9E004669BF72CE2B11D119DA5DD2DE88FB415ABB (17048 bytes)
    C978DD664A96642FEC766BBE7437AA38E45122C5 (17044 bytes)
    BE251F262B4773F8175FFDF8DBF58D43F0EE4E2E (17036 bytes)
    1853340C8B1A9A9FE9BD49087E2BF76D20A9D108 (17046 bytes)
    611E448B9F9A7476DA6D6DE7748707CDC2B5A492 (17046 bytes)
    21CF6BF9E1651EA33524BF6D244638E8021D2C92 (17034 bytes)
    C5D4D89D05180848C97D3032E50F0EF4AF074E69 (17032 bytes)
    FDDB97C91EBD25E5ACC3638F3501A28B3EB0B1B8 (17044 bytes)
    94E4A5EEC33F083B6256C74EA5C7D510ADA9861E (17026 bytes)
    BB8A75274C450E0D9E0914043BB1300AAADEB16C (17032 bytes)
    EC3091656CC4F394BF1C2E9C6C8C8D9B2B22A1EF (17026 bytes)
    ... (259 total)
  }#3E2838E43D52804F9A6AC587F2502FE6FB4927DA
  EvidenceData{
  }#
  Commit{
    BlockID:    7197306281BBB6466B76926C68250195D9BDB014:1:29B22A6A1491
    Precommits: Vote{0:4186A0B94847 14042/00/2(Precommit) 7197306281BB 5F73E273D9BB @ 2018-08-21T06:16:39.480495771Z}
    Vote{1:B2F6015114DF 14042/00/2(Precommit) 7197306281BB F4E075945D4A @ 2018-08-21T06:16:35.451737638Z}
    nil-Vote
    Vote{3:EF8B18476BED 14042/00/2(Precommit) 7197306281BB 3F3B9ABA3CC2 @ 2018-08-21T06:16:39.379867629Z}
  }#50B2D83F6463AD406C28CAF6D1D0AEF6DCCF47D6

```

**Config (you can paste only the changes you've made)**:
max_packet_msg_payload_size = 10240
send_rate = 10000000
recv_rate = 10000000
timeout_precommit = 2000
timeout_prevote = 2000
skip_timeout_commit = true
create_empty_blocks_interval = 30

**node command runtime flags**:

**`/dump_consensus_state` output for consensus bugs**
```
{
                    "round": "38",
                    "prevotes": [
                        "Vote{0:4186A0B94847 14043/38/1(Prevote) 000000000000 1A1112E086A5 @ 2018-08-21T06:33:38.43850798Z}",
                        "Vote{1:B2F6015114DF 14043/38/1(Prevote) 5E351D4F3031 5ACB5AC17EDF @ 2018-08-21T06:33:17.63205982Z}",
                        "nil-Vote",
                        "Vote{3:EF8B18476BED 14043/38/1(Prevote) 000000000000 061726AB3FD6 @ 2018-08-21T06:33:38.342449906Z}"
                    ],
                    "prevotes_bit_array": "BA{4:xx_x} 28/37 = 0.76",
                    "precommits": [
                        "Vote{0:4186A0B94847 14043/38/2(Precommit) 000000000000 073B92F18B31 @ 2018-08-21T06:34:01.807955614Z}",
                        "Vote{1:B2F6015114DF 14043/38/2(Precommit) 000000000000 2469C54395CC @ 2018-08-21T06:34:01.677137679Z}",
                        "nil-Vote",
                        "Vote{3:EF8B18476BED 14043/38/2(Precommit) 000000000000 DC8E17851EEC @ 2018-08-21T06:33:59.544752566Z}"
                    ],
                    "precommits_bit_array": "BA{4:xx_x} 28/37 = 0.76"
                },
                {
                    "round": "39",
                    "prevotes": [
                        "Vote{0:4186A0B94847 14043/39/1(Prevote) 000000000000 1390313DD2B5 @ 2018-08-21T06:34:26.397803427Z}",
                        "Vote{1:B2F6015114DF 14043/39/1(Prevote) 000000000000 0EC23A92A53E @ 2018-08-21T06:34:27.786700519Z}",
                        "nil-Vote",
                        "Vote{3:EF8B18476BED 14043/39/1(Prevote) 000000000000 1B333C16C0EE @ 2018-08-21T06:34:24.522516247Z}"
                    ],
                    "prevotes_bit_array": "BA{4:xx_x} 28/37 = 0.76",
                    "precommits": [
                        "Vote{0:4186A0B94847 14043/39/2(Precommit) 000000000000 2107D35F2F83 @ 2018-08-21T06:34:29.852901395Z}",
                        "Vote{1:B2F6015114DF 14043/39/2(Precommit) 000000000000 F274C2E52CD0 @ 2018-08-21T06:34:27.792260509Z}",
                        "nil-Vote",
                        "Vote{3:EF8B18476BED 14043/39/2(Precommit) 000000000000 49DDE61F9253 @ 2018-08-21T06:34:27.995022481Z}"
                    ],
                    "precommits_bit_array": "BA{4:xx_x} 28/37 = 0.76"
                },
                {
                    "round": "40",
                    "prevotes": [
                        "Vote{0:4186A0B94847 14043/40/1(Prevote) 000000000000 53284452F554 @ 2018-08-21T06:34:53.181048949Z}",
                        "Vote{1:B2F6015114DF 14043/40/1(Prevote) 000000000000 098B8E30C53A @ 2018-08-21T06:34:53.204983703Z}",
                        "nil-Vote",
                        "Vote{3:EF8B18476BED 14043/40/1(Prevote) F665644F5D01 71D8B1CA3B48 @ 2018-08-21T06:34:30.555863498Z}"
                    ],
                    "prevotes_bit_array": "BA{4:xx_x} 28/37 = 0.76",
                    "precommits": [
                        "Vote{0:4186A0B94847 14043/40/2(Precommit) 000000000000 7FBBBB0EB241 @ 2018-08-21T06:35:16.83506436Z}",
                        "Vote{1:B2F6015114DF 14043/40/2(Precommit) 000000000000 DEACAD5CC4B6 @ 2018-08-21T06:35:17.921485105Z}",
                        "nil-Vote",
                        "Vote{3:EF8B18476BED 14043/40/2(Precommit) 000000000000 F23739A23311 @ 2018-08-21T06:35:15.413693698Z}"
                    ],
                    "precommits_bit_array": "BA{4:xx_x} 28/37 = 0.76"
                }
```

**Anything else we need to know**:
I have 4 consensus node. some round, the node received 3 votes, why consensus fail?

**Comment by ebuchman** [comment]
> I have 4 consensus node. some round, the node received 3 votes, why consensus fail?

The votes are for `nil`. From your logs:

```
"Vote{3:EF8B18476BED 14043/40/1(Prevote) F665644F5D01 71D8B1CA3B48 @ 2018-08-21T06:34:30.555863498Z}"
```

This is a vote for block `F665644F5D01`. But this is a vote for nil:

```
"Vote{1:B2F6015114DF 14043/40/1(Prevote) 000000000000 098B8E30C53A @ 2018-08-21T06:34:53.204983703Z}",
```

Note this is different from a `nil-vote`, which just means we didn't receive the vote from that validator.

If we get +2/3 of votes, but not +2/3 votes for a block (ie. some are for nil), then we go to the next round.

Do you have limits set on your block size? It could be related to issues we've seen around https://github.com/tendermint/tendermint/issues/2035

**Comment by xieyiwen** [comment]
thanks for your reply。i understand the difference in voting。

> Do you have limits set on your block size?

yeah, i made max_packet_msg_payload_size = 10240, and the mempool.size is default(=100000).

**Comment by milosevic** [comment]
It could be related to https://github.com/tendermint/tendermint/issues/1745.

**Comment by jaekwon** [comment]
Can you post more logs, especially more of /dump_consensus_state, and perhaps a tar/zip of each of these machine's logs?  It would also help to get snapshots of /net_info for these nodes as this behavior is being observed.

Given that #1745 is about what happens immediately after +2/3 any prevotes are seen, but these validators prevoting nil are waiting some amount of time (~ 15s) and then prevoting nil before seeing +2/3 any, I think something else is at play here, perhaps time to communicate the block parts, or perhaps there's some bug in the way that block parts are being gossipped or reconstructed.

**Comment by ValarDragon** [comment]
I feel like this is related to the maximum block size issues (as bucky mentioned earlier), which are fixed on tendermint v0.24.0.

Can you try to replicate on the latest release, and see if the problem persists? If so, please open a new issue.
